### DISCUSSION TITLE
Add a way to clear eager loaded elements (after field change)

### DISCUSSION THREAD

**Discussion by Anubarak** [craftcms/cms#11003] [Ideas] [answered]
I thought about this many times and always dropped the idea in order to create a custom logic but since I have to implement it in more and more projects I thought it might be time for a feature request.

I would like to have the option to pass `null` into `setEagerLoadedElements` or have a `clearEagerLoadedElements` function.
The reason is the following.. imagine something like this:
```PHP
$entries = Entry::find()
    ->with(['entriesField'])
    ->all();

foreach ($entries as $entry){
    
    // is eager loaded for sure
    if(\is_array($entry->entriesField)){
        echo "initial output eager loaded element:<br>";
        foreach ($entry->entriesField as $relatedEntry){
            // will output Entry B with id 2
            echo "RelatedEntry: {$relatedEntry->title} {$relatedEntry->id}<br>";
        }
    }

    // set the related entry to Entry C with ID 3
    $entry->setFieldValue('entriesField', [3]);

    // save the element... actually this one doesn't matter ¯\_(ツ)_/¯
    Craft::$app->getElements()->saveElement($entry);

    // ....
    // somewhere else in your code in another function
    
    // grab all related elements:
    // since we care about performance, always check if it's eager loaded
    if(\is_array($entry->entriesField)){
        foreach ($entry->entriesField as $relatedEntry){
            // will output Entry B with id 2 -> as before since it's still eager loaded
            echo "RelatedEntry: {$relatedEntry->title} {$relatedEntry->id}<br>";
        }
    }
    
    // BUT if we use the `getFieldValue` way
    echo "via get field value: <br>";
    foreach ($entry->getFieldValue('entriesField')->all() as $relatedEntry){
        // prints Entry C with id 3
        echo "RelatedEntry: {$relatedEntry->title} {$relatedEntry->id}<br>";
    }
}
```
Currently as a workaround I always call:
```PHP
$entry->setEagerLoadedElements('entriesField', $elements);
```
Sometimes it's not that important to actually eager load these elements again after the new field value is set (just need those before). However I don't usually call `->all` at all and use functions like [these ](https://github.com/marionnewlevant/craft-agnostic_fetch/blob/v2/src/helpers/AgnosticFetchHelper.php#L11) so they always give me the wrong value.
By myself in order to be able to use "eager-loading" in other functions despite field changes.
Not sure if this issue is addresses in Craft 4, didn't read all the source code yet but I'll do as soon as I can😅 
Really awesome for Craft 4 would be the following
```PHP
$otherEntry = Entry::findOne(2);
$entry->setFieldValue('relationField', [$otherEntry]);
$entry->relationField // is already eagerLoaded with [$otherEntry]
```

**Comment by brandonkelly** [comment, marked as answer]
I can’t imagine a scenario where it would be expected for an element to retain its eager-loaded elements for a field, _after_ `setFieldValue()` had been called for it. So I’ve just modified `setFieldValue()` to start unsetting any eager-loaded elements for the field being changed, if any, for the next Craft 4 release. (ca0d2dee52fa4abb8afc82700bdcb21d3f176e1a)

**Reply by Anubarak** [reply]
Awesome, that's the best way.👍 
Thank you very much.
I thought there was a special reason I couldn't think of to retain them.

**Reply by brandonkelly** [reply]
Nope, just overlooked :)

**Reply by brandonkelly** [reply]
Craft 4.0.0-RC1 is out now with this change.
